Fixes a few bugs associated with admin-bots section

jajodiaraghav opened the pull request:
The PR resolves the frontend issue associated with #3413 : 

- On deactivating test bot, the edit-bot button gets removed.
- Bot name and email remains striked out if deactivated.
- After reactivating a deactivated bot, the edit-bot button works fine.

Todo : 
- Reactivating a bot should relist it in "Your Bots" section in settings page without reload.
I am not sure how to fix this. It would be great if anyone could help.

smarx commented:
*Automated message from Dropbox CLA bot*

@jajodiaraghav, it looks like you've already signed the Dropbox CLA. Thanks!

HarshitOnGitHub commented:
@jajodiaraghav Zulip uses a real-time sync approach of calling an `zerver/lib/actions.py` function for updating the database related things and sending an event to update open browser windows. Currently the function used for reactivating the bot(`do_reactivate_user`) sends an event only for users. It should be updated to send an event for bots as well.

jajodiaraghav commented:
Thanks a lot @HarshitOnGitHub 😄

jajodiaraghav commented:
@HarshitOnGitHub Done 👍

HarshitOnGitHub commented:
@jajodiaraghav Thanks! Some frontend tests are failing probably due to some class related changes. Please have a look and fix them.

jajodiaraghav commented:
The issue is not with the class here. As we click `reactivate` under bots section, it sends a POST request to `/json/users/{bot@email.com}/reactivate`. It turns out that the bot gets reactivated but it returns `{"msg":"Internal server error","result":"error"}` with a status `500`. How can we fix this?

timabbott commented:
Better debug the 500; there should be a traceback in the server.log.

HarshitOnGitHub commented:
@jajodiaraghav The error is because we are accessing the bot owner's email address even when it is not set, in `notify_created_bot`. We need to conditionally check if the bot has an owner. If the `bot_owner` is None then don't include the owner in the event dict.

HarshitOnGitHub commented on static/templates/admin_user_list.handlebars:
```diff
@@ -34,16 +34,13 @@
   </td>
 </tr>
-{{#if is_active}}
 <tr class="user-name-form display-none" id="user_form_{{email}}">
```
@jajodiaraghav I think this change belongs to the first commit.
I am not sure if this conditional should be removed. @timabbott Do you think it will be fine to remove the `if` conditional from here?

HarshitOnGitHub replied:
Under `Deactivated Users` tab the name and email of users are striked out. It shouldn't be the case

HarshitOnGitHub replied:
On deactivating a user the edit user button should also go away.

jajodiaraghav replied:
I guess that was desired as on removing the `deactivated_user` class, one frontend test fail.

jajodiaraghav replied:
@timabbott Do we need to strike out the deactivated users as we have a front end test which uses the `deactivated_user` class that adds the strike-through style to name and email table data?

timabbott replied:
We did that intentionally to provide an easy way to undo after making changes by accidentally.  Of course, the original implementation didn't have a confirmation modal and so it was not hard to accidentally click the wrong thing; now that we have one, it's possible we no longer need the strikethrough approach.  What do you think?

jajodiaraghav replied:
I personally do think the strikethrough approach is needed as long as we use the confirmation modal. But the frontend test *aaghh* still uses the `deactivated_user` class 😞

jajodiaraghav replied:
@timabbott Should i keep the strikethrough under users tab for now, so that this PR can be merged?

HarshitOnGitHub commented on static/js/admin.js:
```diff
@@ -438,6 +438,7 @@ function _setup_page() {
             success: function () {
                 var button = row.find("button.deactivate");
+                row.find('button.open-user-form').hide();
                 button.addClass("btn-warning");
```
You can make it just `row.find('.open-user-form').hide()`;

jajodiaraghav replied:
With due respect, i don't understand why this is incorrect. `button.something` has been used as jquery selector several times within the file.

HarshitOnGitHub replied:
I am not saying it is wrong. `row.find('.open-user-form').hide();` looks more clean. You can leave it as it is if it matches the style of the rest of the file. :)

HarshitOnGitHub commented on static/js/admin.js:
```diff
@@ -471,6 +472,7 @@ function _setup_page() {
                 row.find(".user-admin-settings").show();
                 var button = row.find("button.reactivate");
+                row.find("button.open-user-form").show();
                 button.addClass("btn-danger");
```
Same here.

HarshitOnGitHub commented on zerver/lib/actions.py:
```diff
@@ -376,7 +376,7 @@ def stream_name(stream):
                           avatar_url=avatar_url(user_profile),
-                          owner=user_profile.bot_owner.email,
+                          owner=user_profile.bot_owner.email if user_profile.bot_owner else None,
                           ))
```
Minor but I think it will be good to have it like this:
```
                            default_all_public_streams=user_profile.default_all_public_streams,
                          avatar_url=avatar_url(user_profile),
                          ))
    # Set the owner key only when the bot has an owner.
    # The deault bots don't have an owner. So don't
    # set the owner key while reactivating them.
    if user_profile.bot_owner is not None:
        event['bot']['owner'] = user_profile.bot_owner.email
    send_event(event, bot_owner_userids(user_profile))
```

HarshitOnGitHub replied:
Add an empty line between your commit message and commit description. This helps github in differentiating between them.

jajodiaraghav replied:
I added a blank line b/w commit message and commit description, but it didn't work for commit: 
`admin.js: Toggle visibility of 'open-user-form' on user/bot deactivation.`

jajodiaraghav commented:
Made some changes. Please review. :)

timabbott commented:
@jajodiaraghav it looks like we already merged a version of the backend part of your branch in f0121973d2ba28b78b542c106f4f92047c76c47c; can you rebase?

timabbott commented:
There's just 4 lines of frontend changes, and I'm having a hard time figuring out how to reproduce the issue they fix.  @jajodiaraghav maybe after rebasing, you can squash your changes and make the commit message clear about the precise bug this fixes?  

With your changes, I can still reproduce deactivating a bot in the /#settings/your-bots page doesn't live-update the /#administration version to show a reactivate button, which is what I thought this was supposed to fix...

jajodiaraghav commented:
@timabbott Updated the commits!

> With your changes, I can still reproduce deactivating a bot in the /#settings/your-bots page doesn't live-update the /#administration version to show a reactivate button, which is what I thought this was supposed to fix.

Although this wasn't mentioned as an issue in #3413, I am trying to fix this although having a hard time figuring out how to. Could you help?

jajodiaraghav commented:
@timabbott Updated commits :)

jajodiaraghav commented:
@timabbott Is the travis error due to master branch?

HarshitOnGitHub commented:
@jajodiaraghav The error is not due to the master branch but rather due to the fact that you have not capitalized the strings correctly. You will need to change "Edit User" to "Edit user". See http://zulip.readthedocs.io/en/latest/translating.html#capitalization for more information.

HarshitOnGitHub commented:
@timabbott Do you think we should change that message to something like "Strings not properly capitalized in frontend" instead of "Strings not capitalized in frontend"?

jajodiaraghav commented:
Thanks for the help @HarshitOnGitHub
I too figured out the error, but it would be nice to update the statement to something more explicit :)

timabbott commented:
Yeah let's improve the error -- that check is new in the last day and I think we could make it a lot clearer.

timabbott commented:
(Just submit a PR!)

HarshitOnGitHub commented:
@timabbott Done #4017.

timabbott commented:
@jajodiaraghav for the second commit, I think you need the update to go in `server_events.js`.  Otherwise, this won't work if there are multiple browser windows open and you do the editing in one of them -- the second won't update.

Also it looks like the first commit has a stray whitespace change; can you fix that?

jajodiaraghav commented:
Done! Although i couldn't get the travis error right. :'(

HarshitOnGitHub commented:
@jajodiaraghav Okay I did some manual debugging and the error is due to the fact that you are having a call to `admin.setup_page()` in the remove bot event but that function is not stubbed in node tests and hence the error. To fix it you will need to stub the `admin.setup_page()` function. However I strongly feel that you should not be calling the `admin.setup_page()` for updating the users table since that is pretty expensive operation just for changing the state of a button. I think you should be modifying the admin's `update_user_data()` function for making such kind of changes. You will probably want to get the relevant bot using the user_id and set the state of the button using the bot's `is_active` flag. Let me know if you get any problem in doing so.
Regarding the irrelevant traceback, I have also been annoyed by it a lot of times but always forgot to open an issue. But this time when I saw it again, the first thing I did was to open an issue for this. It will probably get fixed today. :)

jajodiaraghav commented:
Thanks @HarshitOnGitHub

jajodiaraghav commented:
@HarshitOnGitHub The changes were probably made after i updated my commits.
Anyway, I have updated the code. Thanks for reviewing this. :)

@timabbott

timabbott commented on static/js/admin.js:
```diff
@@ -50,6 +50,30 @@ exports.update_user_data = function (user_id, new_data) {
+            reactivate_button.addClass("deactivate");
+            reactivate_button.removeClass("reactivate");
+            reactivate_button.text(i18n.t("Deactivate"));
+            user_row.removeClass("deactivated_user");
```
This seems to duplicate a lot of code we already have and to be likely very fragile.  

Can you extract (most of) this into a function that can be shared with or replace the code we currently have for managing these objects?  (Or maybe should this replace the existing logic in the `success` handler for the deactivate button?)

jajodiaraghav replied:
Can a do this i a different commit?

timabbott replied:
What I'd personally do is actually rebase and add a first commit that extracts the logic, and then simplify this commit to just need to call the function and do any extra details.    Could be a good rebase practice opportunity for you :)

jajodiaraghav commented:
@timabbott Done. Although I'm not sure what's causing the travis error.

timabbott commented:
Travis problem looks to be some issue with their infrastructure; will probably work if you rebase to have it retry.

timabbott commented:
Looks good, merged, thanks @jajodiaraghav!
